javatests/arcs/android/host/AndroidHostRegistryTest.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.host
+
+import android.content.Context
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import arcs.core.host.ParticleIdentifier
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.Robolectric
+
+@RunWith(AndroidJUnit4::class)
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class AndroidHostRegistryTest {
+    private lateinit var context: Context
+    private lateinit var service: TestReadingExternalHostService
+    private lateinit var hostRegistry: AndroidManifestHostRegistry
+
+    @Before
+    fun setUp() {
+        context = InstrumentationRegistry.getInstrumentation().targetContext
+        service = Robolectric.setupService(TestReadingExternalHostService::class.java)
+        hostRegistry =
+            AndroidManifestHostRegistry(context, { intent -> service.onStartCommand(intent, 0, 0) })
+    }
+
+    @Test
+    fun hostRegistry_availableArcHosts_containsTestArcHost() = runBlockingTest {
+        assertThat(hostRegistry.availableArcHosts()).contains(
+            hostRegistry.adapterFor(TestReadingExternalHostService())
+        )
+    }
+
+    @Test
+    fun hostRegistry_arcHost_canRegisterUnregisterParticle() {
+        runBlocking {
```

Comment:
runBlocking and runBlockingTest do different things. runBlockingTest fails because it skips over delays. I don't know why it fails, but basically runBlockingTest fails with a completion exception like it's not waiting for everything to finish before exiting the test method.

" This job has not completed yet
java.lang.IllegalStateException: 
	at kotlinx.coroutines.JobSupport.getCompletionExceptionOrNull(JobSupport.kt:1128)
"

Reply:
Jason, have you ever had this happen?

Reply:
This is a valid error, it means that you've leaked a Job/CoroutineContext somewhere. 

If you're explicitly not able to designate a parent context for all jobs generated in the test (using `coroutineContext` from the `runBlockingTest` as the parent), that's [sometimes unavoidable] code smell and means there is a potential for memory leaks of coroutine bodies when - in production - a caller attempts to close out its job's children.

Reply:
But based on the description of the differences between runBlocking and runBlockingTest, this doesn't explain why the former succeeds but the latter doesn't? Wouldn't runBlocking also fail with JobCompletionException after a timeout?

Reply:
In any case, I'm removing register/unregister. Most ArcHosts will have their own private/static lists of particles associated with them, and only ProdHost needs an API like this, mostly for compile time static registration from an annotation processor or bootstrap method.